app/models/assignment.rb

```diff
@@ -392,25 +392,71 @@ def add_group(new_group_name = nil)
     Grouping.create(group: group, assignment: self)
   end
 
-  def add_group_api(new_group_name = nil)
-    if new_group_name.nil?
-      if group_name_autogenerated
-        group = self.course.groups.new
-        group.save(validate: false)
-        group.group_name = group.get_autogenerated_group_name
-        group.save
-      else
-        raise 'A group name was not provided'
+  def add_group_api(new_group_name = nil, members = [])
+    members ||= []
+    group = nil
+
+    Group.transaction do
+      group = if new_group_name.nil?
```

Comment:
In general I'd like to avoid this style of variable assignment. Instead, keep the if statement at the outer level and initialize `group` in each branch. This also allows you to delete the `group = nil` line above.